### DISCUSSION TITLE
Support for something like Rust if-let

### DISCUSSION THREAD

**Discussion by tknawara** [scala/scala3#22285] [General Discussion]
Hi team,

I'm very new here, so apologies if I didn't provide enough details. I'm wondering if you have given any thoughts around supporting something like rust [if let](https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html)

The basic idea is to allow for destructuring in if condition and bind everything in the if scope, you can already do that with `match` of course, but I often find myself only interested in a single case of an enum for example while the remaining cases won't need special handling for them.

This is already supported in different scala constructs like for comprehensions and would be nice to do something like

```scala
if Some(v) = myOption then
   ...
```

instead of 

```scala
if myOption.isDefined then
  val Some(v) = myOption
  ....
```

It gets complicated if you want to use v further in the condition though

```scala
if Some(v) = myOption && v.xyz then
  ...
```

**Comment by He-Pin** [comment]
This reminds me https://github.com/dart-lang/language/issues/4205

**Comment by tknawara** [comment]
Similar discussion happened in F# https://github.com/fsharp/fslang-suggestions/issues/705

**Comment by xiaoshihou514** [comment]
Maybe use the function `.collect`?

**Comment by joan38** [comment]
What's the benefit over:
```scala
myOption match
  case Some(v) if v.xyz =>
    ...
  case None =>
```
?
More compact?

**Reply by He-Pin** [reply]
https://google.github.io/comprehensive-rust/pattern-matching/let-control-flow/if-let.html

Maybe save some typing ?

**Reply by arikapt** [reply]
You can chain with normal if statement. Example you parse some logging, you could use "line contains string" with simple case and for more complex you could do if let Some(captured) = regex(line).

**Comment by kyouko-taiga** [comment]
There's been a (pre-)pre-SIP on this topic with a lengthy explanations of the motivations for such a feature: https://contributors.scala-lang.org/t/pre-sip-s-improving-the-expressiveness-of-conditional-control-flow/6488

There are plans to work on a proper SIP in the near future. Any help is welcome obviously.
